src/hooks/useVisibleElements.ts

```diff
@@ -8,8 +8,13 @@ type VisibilityMap = Record<string, boolean>
  */
 export function useVisibleElements<T extends HTMLElement>() {
   const [visible, setVisible] = useState<VisibilityMap>({})
+  const [forceRender, setForceRender] = useState(false)
   const ref = useRef<T>(null)
 
+  const rerender = () => {
+    setForceRender((trigger) => !trigger)
+  }
+
```

Comment:
You can do this with MutationObserver, it will listen for dom changes and force rerender on change.

No need to pass rerender function to other components.

```suggestion
  useEffect(() => {
    const mutations = new MutationObserver((mutations) => {
      let shouldRerender = false
      mutations.forEach((mutation) => {
        if (mutation.type === "childList") {
          shouldRerender = true
        }
      })

      setForceRender(shouldRerender)
    })

    if (ref.current) {
      mutations.observe(ref.current, {
        childList: true,
      })
    }

    return () => {
      mutations.disconnect()
    }
  }, [])
```